[Merged by Bors] - API phase 9: finish implementing grpc services

lrettig opened the pull request:
## Motivation
See #1764 
Also Fixes #2116 
Requires #2092, merge that first

## Changes
- Adds stubs for all remaining services and endpoints
- Upgrades to latest API build, renames a few API data items to be consistent with improved naming in the API code
- Allows MeshService to read and return layer root hash: makes TransactionProcessor GetLayerStateRoot method public
- Adds a config param (internal, only for testing) to force the syncer to always enable listening to gossip
- Event Reporter no longer reports any details about Status updates, it just "pings" the listener which is responsible for re-fetching all status items they care about
- Adds account, reward, receipt channels to Reporter; allows Reporter to block, or to increase channel buffer sizes, on initialization
- Allows Reporter report methods to block (useful for tests)
- Implements all endpoints that can be implemented (some, e.g. those involving transaction receipt, cannot yet be implemented)
- Adds a method `GetSmesherID` to ATX to allow the nodeID that created it to be read, as part of finding ATXs by smesher for the API

## Test Plan
- Adds tests for all new endpoints and services

lrettig committed:
Implement MeshService.CurrentEpoch

Pass layersPerEpoch into the MeshService

lrettig committed:
Implement and add tests for 3 more endpoints

lrettig committed:
Slightly refactor tests

lrettig committed:
Move max tx per block into config

lrettig committed:
Implement MaxTransactionsPerSecond

and adds test for it

lrettig committed:
WIP implementing LayersQuery

Still need to wrangle some more data

lrettig committed:
Complete, untested implementation of LayersQuery

lrettig committed:
Trivial version of test of new endpoint passing

lrettig committed:
Implement more comprehensive test

Test all return data, test is passing
Fixed a couple of implementation bugs spotted while testing

First E2E working version of LayersQuery

lrettig committed:
Make lint happy

And add some fun comments

lrettig committed:
Handling and testing of errors

Add much more comprehensive testing of errors

Simplify how we handle errors involving missing layers for now

lrettig committed:
Add test for missing inputs

lrettig committed:
WIP: implement AccountMeshDataQuery

lrettig committed:
Add mock mempool

lrettig committed:
Overhaul tests

Clean up considerably, add a bunch more tests for AccountData

lrettig committed:
Run grpc server on localhost for test

This prevents obnoxious macOS popup

lrettig committed:
Pass txpool to new MeshService

lrettig committed:
Make lint happy

lrettig committed:
Basic E2E support for first stream

Includes a basic test

lrettig committed:
Add activations stream and test for same

lrettig committed:
WIP: sending events

service does not read filter yet

lrettig committed:
Properly handle filter in AccountMeshDataStream

lrettig committed:
Make tests much more robust

Add lots more tests for AccountMeshDataStream

Streamline the comprehensive test with another waitgroup

Fix a test issue where order of ATXs was changing due to their being
stored in a map

lrettig committed:
WIP cleaning up comprehensive stream test

lrettig committed:
Lots of cleanup

Broke out complex concurrency test that was working before, but stopped
working after sharing some data structures with the other tests I added.
It's working again as its own standalone test.

lrettig committed:
More test cleanup

- Re-add test removed by accident
- Remove test I meant to remove
- Remove commented-out test for closing stream from client side - there
isn't really any way to test this, and it doesn't seem important to test

lrettig committed:
Streamline all events reporting

Have everything flow through new streamer interface

lrettig committed:
Basic LayerStream implementation

Adds a basic test, which is passing

lrettig committed:
Add todo note about stream context

lrettig committed:
Minor test cleanup

lrettig committed:
Fully validate received layer from LayerStream

Catch an uncaught error in test
Tests for MeshService are finished

lrettig committed:
Move test function

From mesh to types, where it belongs, to avoid a dependency cycle

lrettig committed:
Rename streamer -> reporter

The name is a bit clearer and reflects its general purpose better
Add a non-API related test for the new reporting interface
Make event reporting non-blocking

lrettig committed:
Make linter happy

lrettig committed:
Very minor cleanup

Made a couple of TODO items clearer

lrettig committed:
Subsume pubsub init into new reporter

lrettig committed:
Add config for all services

Also adds a stub implementation of the GlobalStateService

lrettig committed:
Finish GlobalStateService mocks

Pass all enable service flags into JSON server
Update some description comments

lrettig committed:
Finish TxService, SmesherService stubs

lrettig committed:
Some minor cleanup

Remove err retval from Cleanup since it's unused

lrettig committed:
Add error stream

For now just hook into logs to catch error/panic

lrettig committed:
Set up Status channel infrastructure

Rename a bunch of things

lrettig committed:
Subscribe to all relevant status events

lrettig committed:
Hook into zapcore logging to report errors

This prevents a dependency loop from the log package

Make all reporter channel writes nonblocking

Add a basic test for reporting errors

lrettig committed:
Add basic test for ReportNodeStatus

lrettig committed:
WIP writing a node test for the new streams

Both streams are working. Clean up the test to check the contents of the
received messages.

lrettig committed:
Bugfix: reporter must subscribe in a goroutine

lrettig committed:
Cleanup comments

lrettig committed:
Improve NodeService stream tests

lrettig committed:
Add cmdline flag for tcp interface

This, in combination with --acquire-port=false, prevents the obnoxious
warning popup in macOS.

lrettig committed:
Add cmdline flag for grpc server interface

Make this fully specifiable. In combination with previous commit, this
fully works around obnoxious macOS warning popup.

Include this in node test.

lrettig committed:
Add channel to notify that app is started

Used in tests rather than time.Sleep, which is messy. Ideally we'd have
a full set of notification channels, one per service, so that tests
could subscribe to the services they're interested in, but this works
well enough as a stopgap for testing.

lrettig committed:
Improve test

Improve concurrency handling, check more details

lrettig committed:
More sophisticated handling of errors

Make the error channel in reporter buffered, and add a comment
explaining why

Test panicking in the test. Make sure each error is received in the
expected order.

lrettig committed:
Get state root hash for all layers

lrettig committed:
Update to latest api build

lrettig committed:
Remove two unused interface methods

lrettig committed:
Implement GlobalStateHash

Includes basic test

lrettig committed:
Implement and test Account endpoint

Includes test

Clean up method param names from API interface

lrettig committed:
Implement AccountMeshDataQuery MaxResults, Offset

lrettig committed:
Minor cleanup of filters in MeshService

lrettig committed:
WIP: partially implemented AccountDataQuery

There is currently no way to get smesher for each reward

lrettig committed:
Update comment

Add link to issue

lrettig committed:
Bugfix

spotter: @antonlerner

lrettig committed:
Handle some lingering TODO items

Fix small issues discussed with Anton in phase7 review
Sort out layer statuses for LayersQuery and LayerStream
Report layer data more often from Mesh
Encapsulate layer status inside new layers reported using the Reporter

lrettig committed:
Update txProcessor interface in test

To match changes several commits ago

lrettig committed:
Update and clean up tests

To match latest code changes.
Add some explanatory comments.
Fix a bug in Reporter.

lrettig committed:
Finish implementing AccountDataQuery

Adds basic test

lrettig committed:
Mark SmesherDataQuery as unimplemented

for now - we are waiting on #2068

lrettig committed:
WIP implementing AccountDataStream

lrettig committed:
Finish AccountDataStream

Add test
Add more streams to Reporter

lrettig committed:
Bugfix in Reporter

Spotter: @AndrewAR2

lrettig committed:
Update modules after rebase

lrettig committed:
Add blocking option to Reporter

Use it to remove time.Sleep from test

lrettig committed:
Update to latest api module version

Renames a few things

lrettig committed:
Update api module again

lrettig committed:
Rename to match latest API module

lrettig committed:
Finish GlobalStateService, add basic test

lrettig committed:
Add comprehensive test for same

lrettig committed:
Revert "Remove two unused interface methods"

This reverts commit a6a6b5141d58d520bd693683cf3cbce6bae655d2.

lrettig committed:
Update to latest API module

lrettig committed:
Rename things to match latest API build

lrettig committed:
Finish alpha implementation of SmesherService

Lots of endpoints still to be filled in

lrettig committed:
Add basic SmesherService tests and stubs

lrettig committed:
Properly handle input errors

lrettig committed:
Add more tests for unimplemented endpoints

lrettig committed:
Minor cleanup

Cleanup some unused properties in several of the services
Clean up some comments

lrettig committed:
Initial complete implementation of TxService

Adds a test for SubmitTransaction only so far
Clean up the test code a bit
Modify Reporter transaction channel to include tx state
Report invalidated txs from tx pool

lrettig committed:
More comprehensive tests

lrettig committed:
Finish adding TxService tests

lrettig committed:
Add another more robust stream test

lrettig committed:
Update mocked object to match interface change

See c373c24bc94ccf339586d5c6e685e9efa7a3482b

lrettig committed:
Do not allow reporter to be reinitialized

Unless closed first

lrettig committed:
Update service arguments

To match cleanup of their constructors

lrettig committed:
Add AlwaysListen config

Forces syncer to always listen to gossip, useful for testing

lrettig committed:
Add E2E test of tx service

lrettig committed:
Add more bad input tests

lrettig committed:
Tidy modules after rebase

lrettig committed:
Make status reporting much simpler

Don't convey any data as part of the reporting. Let the listener pull
down all of the data. This makes everything simpler and more
streamlined. Update tests accordingly.

lrettig committed:
Add issue links for TODO items

lrettig committed:
Make linter happy

lrettig committed:
Make linter great again

lrettig committed:
Fix test

lrettig committed:
Fix test for CI and Docker

Ignore an error that only happens when there's an existing database

lrettig committed:
Small tweaks to fix tests

lrettig committed:
Skip two defunct tests for now

They will be re-enabled in phase9. The required fixes are too complex to
backport.

lrettig committed:
Register all services to gateway server

lrettig committed:
Register all services before starting grpc server

This mostly worked before due to luck, but this is the order things
should happen in. Make sure they happen in the right order.

lrettig committed:
Make tests less fragile

Make sure global memory is cleared before the NodeService test, which
watches the error stream. Add a sleep statement since we have no way of
ensuring th service is actually listening and this keeps causing issues
in docker/travis.

lrettig committed:
Try fixing the leaking another way

The previous fix did not work because the log itself is a global
singleton. Try another workaround until we have a more robust solution.

antonlerner commented on api/config/config.go:
```diff
@@ -23,9 +47,49 @@ func init() {
+		StartMeshService:        defaultStartMeshService,
+		StartGlobalStateService: defaultStartGlobalStateService,
+		StartTransactionService: defaultStartTransactionService,
+		StartSmesherService:     defaultStartSmesherService,
```
what is the smesher service? perhaps choose a name actually describing the service

lrettig replied:
It lets people who are smeshing configure their allocated space and coinbase, start and stop smeshing, etc. - can you think of a better name? 🤓

antonlerner replied:
Sinc ethis is done for reward or block creation - either rewardSettings or BlockCreationParam?

lrettig replied:
I think the name is okay, let me know if you feel strongly otherwise

antonlerner commented on api/grpcserver/mesh_service.go:
```diff
@@ -0,0 +1,514 @@
+	for _, b := range layer.Blocks() {
+		txs, missing := s.Mesh.GetTransactions(b.TxIDs)
+		// TODO: Do we ever expect txs to be missing here?
+		// E.g., if this node has not synced/received them yet.
```
IMO, some services shouldn't be available until we are synced

lrettig replied:
Let's follow up here... https://github.com/spacemeshos/api/issues/100

antonlerner commented on api/grpcserver/mesh_service.go:
```diff
@@ -0,0 +1,514 @@
+
+		var pbTxs []*pb.Transaction
+		for _, t := range txs {
+			if t == nil {
```
is this supposed to happen?

lrettig replied:
Good catch. I don't think this should happen. And I'm not doing this check elsewhere. I'll remove it. My guess is that I probably added it because there was a funky test case that accidentally returned a null tx at some point.

antonlerner commented on api/grpcserver/mesh_service.go:
```diff
@@ -0,0 +1,514 @@
+		for _, t := range txs {
+			if t == nil {
+				log.Error("got empty transaction data from layer %v", layer.Index())
+				return nil, status.Errorf(codes.Internal, "error retrieving tx data")
```
you can send the original message in the error as well, maybe it will help debug issues faster later

lrettig replied:
My strategy here has been to not expose the underlying error to the API caller. Instead, I return a generic error message to the caller, and log the full error. The logic here has to do with security... it's probably not a super strong argument, but it seems better if someone hitting the API doesn't have too much information about what's going on under the hood. LMK if you disagree, we can discuss.

antonlerner replied:
for now I'd prefer more info than less, since we will be the first customers of both node and wallet

lrettig replied:
https://github.com/spacemeshos/go-spacemesh/issues/2137

antonlerner commented on api/grpcserver/mesh_service.go:
```diff
@@ -0,0 +1,514 @@
+		if l <= lastLayerPassedHare.Uint64() {
+			layerStatus = pb.Layer_LAYER_STATUS_APPROVED
+		}
+		if l <= lastLayerPassedTortoise.Uint64() {
```
won't this cause all layers to have a confirmed status - even if tortoise haven't confirmed them yet?

lrettig replied:
Hmm. I reread this logic and I think it's correct. If a layer has passed neither Tortoise nor Hare, it's marked UNSPECIFIED. If it's passed only Hare, it's marked APPROVED. If it's passed only Tortoise, or both, it's marked CONFIRMED. Is there an error in this logic?

Please make sure you double-check lines 348-349 above, too.

lrettig replied:
BTW the logic for this was determined here: https://github.com/spacemeshos/api/issues/89#issuecomment-655625229

antonlerner replied:
I meant that `s.Mesh.ProcessedLayer()` is the layer last pushed to tortoise, the tortoise then outputs newPbase which is the actual layer that was confirmed

lrettig replied:
https://github.com/spacemeshos/go-spacemesh/issues/2138

antonlerner commented on api/grpcserver/mesh_service.go:
```diff
@@ -0,0 +1,514 @@
+			if activation.Coinbase == addr {
+				pbActivation, err := convertActivation(activation)
+				if err != nil {
+					log.Error("error serializing activation data: ", err)
```
you can reuse error msg

antonlerner commented on api/grpcserver/node_service.go:
```diff
@@ -0,0 +1,181 @@
+// STREAMS
+
+// StatusStream exposes a stream of node status updates
+func (s NodeService) StatusStream(request *pb.StatusStreamRequest, stream pb.NodeService_StatusStreamServer) error {
```
should status be a stream or query?

lrettig replied:
It's designed to be a stream. Queries are for historical data (which sort of doesn't make sense for status). With a stream, the client can always know immediately when the realtime status changes.

antonlerner commented on api/grpcserver/transaction_service.go:
```diff
@@ -0,0 +1,267 @@
+		return nil, status.Errorf(codes.InvalidArgument,
+			"`Transaction` must contain a valid, serialized transaction")
+	}
+	log.Info("GRPC TransactionService.SubmitTransaction to address: %s (len: %v), "+
```
do we use gas limit?

lrettig replied:
The transaction type has a field called `GasLimit`. Not sure if it's currently used or not. It's implemented in the existing API, and used in tests, etc.

antonlerner commented on api/grpcserver/mesh_service.go:
```diff
@@ -0,0 +1,514 @@
+		// but to read all of them.
+		// TODO: index activations by layer (and maybe by coinbase)
+		// See https://github.com/spacemeshos/go-spacemesh/issues/2064.
+		var activations []types.ATXID
```
I'd at least refactor to separate functions to avoid long monolithic functions, IMO

antonlerner commented on api/grpcserver/smesher_service.go:
```diff
@@ -0,0 +1,155 @@
+		return nil, status.Errorf(codes.InvalidArgument, "`Id` must be provided")
+	}
+
+	addr := types.BytesToAddress(in.Id.Address)
```
should we use the function that returns error? to make sure we don't receive invalid input, this function converts anything so it might be a bit dangerous

lrettig replied:
Hmm. The code is dead simple. Do you think there's a potential issue here? I mean, literally any 20 bytes is a valid possible address. The only thing we could check is the length (which I think is checked implicitly in `SetBytes`)

```
func BytesToAddress(b []byte) Address {
	var a Address
	a.SetBytes(b)
	return a
}
```

antonlerner commented on events/reporter.go:
```diff
@@ -0,0 +1,430 @@
+		Valid:       valid,
+	}
+	if reporter != nil {
+		if reporter.blocking {
```
should this be thread safe? accessing reporter properties

lrettig replied:
The properties are read only so I think we're good here, but you're right re: thread safety and setting the singleton reporter in the first place. In production code there should only ever be a single instance and it should only ever be initialized/torn down once, but it can't hurt to make it thread safe just in case -- will fix this.

lrettig replied:
how is this? https://github.com/spacemeshos/go-spacemesh/pull/2071/commits/8a94e2dd8214359d182a9f2ceb94b87dd8725917

antonlerner commented on events/reporter.go:
```diff
@@ -0,0 +1,430 @@
+// new layer is reached.
+func SubscribeToLayers(newLayerCh timesync.LayerTimer) {
+	if reporter != nil {
+		// This will block, so run in a goroutine
```
who calls this function? can it leak go routines?

lrettig replied:
Good question. `newLayerCh` comes from `clock.Subscribe()` which uses a non-blocking send. This service is set up the same way as all of the other services which subscribe to the layer ticker. I think this code is okay. LMK if I'm missing anything.

antonlerner commented on events/reporter.go:
```diff
@@ -0,0 +1,430 @@
+
+// ReportNewLayer reports a new layer
+func ReportNewLayer(layer NewLayer) {
+	if reporter != nil {
```
you can just register to the appropriate pubsub channel within the node to receive these events, 
otherwise, why keep both?

lrettig replied:
The reporter events and the pubsub events are pretty different. For one thing, pubsub events are all serialized and sent over TCP, whereas the reporter events are native types and kept in-process. I talked more about this here: https://github.com/spacemeshos/SMIPS/issues/21#issuecomment-653199068. Let's discuss if you still have concerns!

antonlerner reviewed (changes requested):
see comments and fix lint

lrettig committed:
Make lint happy

Clean up tests
Rename a GRPC method to make lint happy

lrettig committed:
Fix merge conflicts, update API version

Update naming to match latest API build
Finish resolving merge conflicts from rebase on phase8

lrettig committed:
Always close reporter before reinit in tests

Should prevent error about already being initialized

lrettig committed:
Re-enable and fix failing tests

Log the number of registered services from grpc server

lrettig committed:
Bugfix: run http server in a goroutine

Same as grpc server, since this blocks

lrettig committed:
Bugfix: make sure context isn't closed too early

Run the entire surrounding method in a goroutine

lrettig committed:
Minor improvements to tests

Make sure the event reporter is closed between tests

lrettig committed:
Make test less flakey

Adding a brief sleep (womp womp) seems to prevent sporadic test failures
due to listener not listening yet

lrettig committed:
Small fix to test

lrettig committed:
Make thread-safe

lrettig committed:
Remove an unnecessary check

We don't perform this check elsewhere. It shouldn't be necessary. Per
@antonlerner in review.

lrettig committed:
Fix test

lrettig committed:
Improve concurrency in test

lrettig committed:
Better isolate tests

The global Context was causing one test to cancel another. Protect the
tests with a setup method that isolates them, and have each test use a
different port to isolate them even more.

lrettig committed:
Remove unused args from tx service

lrettig committed:
Small tweaks to improve logging

Print full error message from mesh
Pretty-print layer info in reporter logs

lrettig committed:
Remove a redundant report

lrettig committed:
WIP troubleshoot LayerStream issue

See #2084

lrettig committed:
Fix error about missing layer state root

This is expected. Update log message accordingly.

lrettig committed:
Fix debug message

lrettig committed:
Add link to issue for TODO

lrettig committed:
Fix lingering merge conflict

Not sure how I missed this!

lrettig committed:
Bugfix: correctly calculate CommitmentSize

This refers to the amount of space being committed to, not to the size of the ATX itself!

lrettig committed:
Make linter happy, fix a test

Don't try to predict in the test what happens after the app closes and
the stream closes, this appears to be nondeterministic

lrettig committed:
Fix another test

lrettig committed:
Fix test

lrettig committed:
Bugfix, add debug output

Catch empty input before it causes a null pointer exception

lrettig committed:
Final cleanup, per review

per Anton's review
reuse error message
factor out some code to make parent method shorter

lrettig committed:
More cleanup post refactor

lrettig committed:
One more for the refactor

Fix https://github.com/spacemeshos/go-spacemesh/issues/2116

lrettig committed:
Fix more imports post-rebase

lrettig committed:
Fix tests post-rebase

Check for nil pointers

lrettig commented:
bors try

lrettig commented:
bors try

antonlerner commented on api/grpcserver/globalstate_service.go:
```diff
@@ -0,0 +1,457 @@
+	}}, nil
+}
+
+// Account returns data for one account
```
`\\ Account returns Balance and tx counter for one account`

antonlerner commented on api/grpcserver/globalstate_service.go:
```diff
@@ -0,0 +1,457 @@
+			// nonce.
+			balance := s.Mesh.GetBalance(updatedAccount)
+			counter := s.Mesh.GetNonce(updatedAccount)
+			if err := stream.Send(&pb.GlobalStateStreamResponse{Datum: &pb.GlobalStateData{Datum: &pb.GlobalStateData_AccountWrapper{
```
I'd create a constructor method for these PB types for bette readability

lrettig replied:
Good point. I agree that would be better. There are dozens of them, and that's a pretty big project. Are you okay if I open another task for this and leave it as an improvement?

lrettig replied:
#2118

lrettig committed:
Make endpoint function description clearer

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
An assertion error in `conftest.py` keeps failing when I run bors:

```
        if not match:
>           raise Exception("Failed to read container logs in {0}".format('bootstrap'))
E           Exception: Failed to read container logs in bootstrap
setup_utils.py:152: Exception
```

Not sure what's going on. I don't think this branch changes the system tests config at all. I tried running bors on another, unrelated branch and it worked. Could possibly be related to #2090. Will investigate.

moshababo commented:
You also need to integrate latest updates from `develop`, mainly https://github.com/spacemeshos/go-spacemesh/commit/efa7538ba4fbb4ea90be43c827233a580c736e19, no?

lrettig commented:
> You also need to integrate latest updates from `develop`, mainly [efa7538](https://github.com/spacemeshos/go-spacemesh/commit/efa7538ba4fbb4ea90be43c827233a580c736e19), no?

Already done, this branch was rebased on that one!

lrettig committed:
Fix small typo

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try-

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try-

lrettig commented:
bors try

avive commented:
bors try

avive commented:
> ## try
> Build failed:
> 
> * [continuous-integration/travis-ci/push](https://travis-ci.org/github/spacemeshos/go spacemesh/builds/720545613?utm_source=github_status&utm_medium=notification)

@lrettig - looks like you are getting travis status via bors now again....

lrettig commented:
bors try

lrettig commented:
bors try-

lrettig commented:
This branch should be ready to go, except that it needs the fix from #2092 for the logging issue. I want to merge that into develop first. (That's what #2119 was, and it's passing all of the tests.)

lrettig commented:
bors try

lrettig committed:
Merge branch 'develop' into api-phase9

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors merge
